src/mlpack/methods/decision_tree/best_binary_categorical_split_impl.hpp

```diff
@@ -0,0 +1,414 @@
+  // to right tree Tᵣ. 
+  else 
+  {
+
+    // A map from category Cⱼ to the samples whose categorical value
+    // for variable vₖ is Cⱼ. The jth column corresponds to Cⱼ. 
+    SpMat<short> categorySamples(n, numCategories);
+    for (size_t i = 0; i < n; ++i)
+      categorySamples(i, data[i]) = 1;
+
+    uvec categories(numCategories); 
+    if (UseWeights)
+    {
+        // Weight counts for computing the gain. 
+        double totalWeight = accu(weights);
+        mat classWeightSums = zeros<mat>(numClasses, 2);
+        bestFoundGain *= totalWeight;
+
+        // Recursively check the gain for all partitions.
+        improved = PartitionSplit(
+            data, labels, numCategories, numClasses, weights, 
+            totalWeight, bestFoundGain, categorySamples, categories, 
+            splitInfo, classWeightSums
+        );
+        bestFoundGain /= totalWeight;
+    }
+    else
+    {
+        // Class counts for computing the gain. 
+        Mat<size_t> classCounts = zeros<Mat<size_t>>(numClasses, 2);
+        bestFoundGain *= n; 
+
+        // Recursively check the gain for all partitions.
+        improved = PartitionSplit(
+            data, labels, numCategories, numClasses, bestFoundGain, 
+            categorySamples, categories, splitInfo, classCounts
+        );
+        bestFoundGain /= n; 
+    }
+  }
+  return improved? bestFoundGain : DBL_MAX;
+}
+
+
+// Overload used in regression with MSEGain.
+template<typename FitnessFunction>
+template<bool UseWeights, typename VecType, typename ResponsesType,
+         typename WeightVecType>
+double BestBinaryCategoricalSplit<FitnessFunction>::SplitIfBetter(
+    const double bestGain,
+    const VecType& data,
+    const size_t numCategories,
+    const ResponsesType& responses,
+    const WeightVecType& weights,
+    const size_t minLeafSize,
+    const double minGainSplit,
+    vec& splitInfo,
+    AuxiliarySplitInfo& aux,
+    FitnessFunction& fitnessFunction)
+{
+  static_assert(std::is_same<FitnessFunction, MSEGain>::value,
+      "BestBinaryCategoricalSplit: regression FitnessFunction must be MSEGain."
+  );
+  const size_t n = data.n_elem;
+  double bestFoundGain = std::min(bestGain + minGainSplit, 0.0);
+
+  // We are too small to split.
+  if (n < (minLeafSize * 2))
+    return DBL_MAX;
+
+  // Order the categories of variable vₖ by increasing mean
+  // of the response y. categoryResponse[i, 0] will contain
+  // the mean response for category Cᵢ.
+  vec categoryResponse(numCategories, fill::zeros);
+  uvec categoryCounts(numCategories, fill::zeros);
+
+  for (size_t i = 0; i < n; ++i)
+  {
+      categoryResponse[data[i]] += responses[i];
+      ++categoryCounts[data[i]]; 
+  }
+  for (size_t i = 0; i < numCategories; ++i)
+      categoryResponse[i] =  categoryCounts[i] == 0 ? 0 : 
+          categoryResponse[i]/categoryCounts[i];
+
+  uvec sortedCategories = sort_index(categoryResponse);
+  uvec categoryRank(numCategories);
```

Comment:
Ok, I will see what I can do.

Reply:
Let me know if the suggestion above doesn't work, I can try again if you like. :+1:

Reply:
@rcurtin The Armadillo one-liners did not work for me (yet): I am getting some arcane template-related output from the GCC compiler. For the preceding one-liner, I get:
```
error: no match for ‘operator[]’ (operand types are ‘arma::u
vec’ {aka ‘arma::Col<long long unsigned int>’} and ‘arma::Col<long long unsigned int>’)    
   67 |       arma::uvec transformedData = categoryRank[arma::conv_to<arma::uvec>::from(data)];
```
followed by a long list of failed substitutions. The Armadillo library and its types are still relatively foreign to me: before I dig in, if it's simple for you, I would definitely appreciate a suggestion here. Then I did not find an elegant way to simplify lines 183-186. If you have anything particular in mind, please let me know.

Reply:
I fought with it for a little while, but I came to the same conclusion as you.  Anyway, I suppose it's fine as-is; maybe someday later it can be cleaned up.